Call the Shopify on the Solidus Update / Destroy / Create for a product

thisiscab opened the pull request:
Problematic case:

Once a Solidus gets updated, we want to trigger an after_update event
that will trigger the task that will update this product to Shopify.

When we update it on Shopify, in a certain case the product on Shopify
may have been deleted so we need to re-create it. When we re-create it,
we need to save the shopify_product_id to the Solidus product. And there
goes an infinite loop

This solves that.

thisiscab committed:
Update the shopify id field without triggering the update callback

Problematic case:

Once a Solidus gets updated, we want to trigger an after_update event
that will trigger the task that will update this product to Shopify.

When we update it on Shopify, in a certain case the product on Shopify
may have been deleted so we need to re-create it. When we re-create it,
we need to save the shopify_product_id to the Solidus product. And there
goes an infinite loop

This solves that.

thisiscab committed:
Add the callback events to propagate product data to Shopify

We want to make sure that we correctly save the Solidus product on
Shopify when the product gets edited / destroyed / updated.

This ensure that the Solidus products are 1:1 with Shopify

thisiscab commented on app/models/spree/product_decorator.rb:
```diff
@@ -0,0 +1,35 @@
+      if assembly?
+        service = Shopify::BundledProductExporter.new(spree_product: self)
+      else
+        service = Shopify::ProductUpdater.new(spree_product_id: id)
```
We probably have to change the initializer interface here

thisiscab commented on app/models/spree/product_decorator.rb:
```diff
@@ -0,0 +1,35 @@
+    end
+
+    def update_shopify_product
+      if assembly?
```
I am unsure if this should be done in some kind of factories, what do you guys think?

thisiscab commented:
Whenever you guys are ready! 
@DanielWright @braidn @bryanmtl 💃

braidn commented:
This is much more manageable. Thank you.

thisiscab commented:
Damnnn, that implementation can't work. Since I'm creating Solidus objects in my specs to test this feature. So when I create a Solidus object in my specs, it directly exports them to Shopify, which makes all kind of funky results.

Do you guys have suggestions?

I could go in the admin controller, decorate it so that it does it only there...

DanielWright commented:
Can you make callbacks conditional? Maybe implement a predicate method (e.g. `should_export_to_shopify?`) that encapsulates the conditions under which a product should or should not be sent to Shopify?

thisiscab commented:
That would have been a good idea, but I can't think of a condition where we wouldn't want to save it to Shopify :/

We could have a flag that is turned on or off depending on some scenario, but that seems funky to me.

DanielWright commented:
If the situation is you don't want _your tests_ to export your products to Shopify, then why not adjust your test environment to stub that behaviour? A tool like WebMock would allow you to stub requests to Shopify such that your object think it's talking to Shopify, but the messages are going nowhere.

thisiscab committed:
Add a "do not export to shopify" flag on a product

There are some scenarios where we may not want a solidus product to be
updated onto Solidus. Having that attr_accessor flag will let us answer
those cases.

Also, this is really useful to us in the sense that we do not want the
product to be already exported on Shopify when we want to test the
product exporter. In order to run our spec, we are creating solidus
object and previously when we would create those object, they would
already be exported to Shopify.

braidn commented on app/models/spree/product_decorator.rb:
```diff
@@ -0,0 +1,41 @@
+module Spree
+  module ProductDecorator
```
Is this something we want to decorate onto the Product class? Or should we build some kind of factory class that encapsulates all the magic?

braidn commented on app/services/shopify/association_saver.rb:
```diff
@@ -11,16 +11,14 @@ def save_pos_variant_id_for_variants(spree_variants, shopify_variants)
       def save_pos_variant_id(spree_variant, shopify_variant)
-        spree_variant.pos_variant_id = shopify_variant.id
-        spree_variant.save
+        spree_variant.update_columns(pos_variant_id: shopify_variant.id)
```
I know there was a question about how this doesn't initiate callbacks this morning. However, I read these methods like:

`we are updating the column directly in the db and bypassing the model completely`

This is why I think it reads fine and clues the reader into what's 'sort of' going on.

braidn commented on spec/models/spree/product_spec.rb:
```diff
@@ -0,0 +1,88 @@
+require 'spec_helper'
+
+module Spree
+  RSpec.describe Product do
```
Yeah reading the decorator and this, my twitch would be to create a class/object to encapsulate this work that's not the product itself. This also helps with removing God objects which tend to rear their heads on User/Product/Order. Totally up to you though.

braidn reviewed (commented):
MUCH better Cab. Easliy able to look through and lay 👀 on the all of the code. Thank you again.

thisiscab commented:
I'd like to touch base about the feature that this adds here. I am hitting so many problems with that implementation, where it sends the data to shopify on callbacks. Having a verbal discussion would help me :)

braidn commented:
Write away! Or... did you mean phone/video?

thisiscab commented:
Phone

thisiscab commented:
I would like to continue the conversation here @braidn https://github.com/glossier/solidus_retail/pull/34/files

thisiscab commented:
We have hit many problems with the way callback works. It was really hard to make sure that under certain scenario stuff wouldn't save to Shopify but in other they would. 

Instead we have decided to implement those calls directly in the action of the Admin Controller.

thisiscab commented:
We have hit many problems with the way callback works. It was really hard to make sure that under certain scenario stuff wouldn't save to Shopify but in other they would. 

Instead we have decided to implement those calls directly in the action of the Admin Controller.
